### ISSUE TITLE
2.2.0: raw module regression under sudo (with patch)

### ISSUE THREAD

**Issue by dch** [ansible/ansible#16311]
##### ISSUE TYPE
- Bug Report
##### VERSION

devel at https://github.com/ansible/ansible/commit/624e7dae544292740a691e369a58320c21d67537

```
ansible 2.2.0
  config file = 
  configured module search path = Default w/o overrides
```
##### OS / ENVIRONMENT
- OSX master
- FreeBSD 10.3R amd64 nodes
##### SUMMARY

greedy parsing in raw module consumes parameters and not just executable task
##### STEPS TO REPRODUCE

``` yaml
# site.yml

---
- name: deploy and configure site
  hosts: all
  become: yes
  gather_facts: yes
  vars_files:
    - vars/users.yml
  roles:
    - pkg
...
```

``` yaml
# roles/pkg/tasks/main.yml
# ensures pkg(8) is bootstrapped, using correct repos and up to date
# install python in raw mode for ansible

---
- name: check if pkg tool needs bootstrapping
  raw: /usr/sbin/pkg -N
  # executable=/bin/sh commented out but recommended by @bcoca
  register: pkg
  ignore_errors: True
  tags:
    - pkg
    - bootstrap

- name: bootstrap pkg tool
  raw: /usr/bin/env ASSUME_ALWAYS_YES=1 /usr/sbin/pkg bootstrap -f
  when: pkg|failed
  tags:
    - pkg
    - bootstrap

- name: force clean pkg repositories
  raw: /usr/bin/env ASSUME_ALWAYS_YES=1 /usr/sbin/pkg clean -a
  when: pkg|failed
  tags:
    - pkg
    - bootstrap

- name: retrieve latest pkg indexes
  raw: /usr/bin/env ASSUME_ALWAYS_YES=1 /usr/sbin/pkg update -f
  when: pkg|failed
  tags:
    - pkg
    - bootstrap

- name: install python so that ansible can run in non-raw mode
  raw: /usr/sbin/pkg install -y python2
  when: pkg|failed
  tags:
    - pkg
    - bootstrap
```
##### EXPECTED RESULTS

from ansible 2.1.0:

``` shell
 ansible-playbook site.yml --limit mrbrown\* -vvv
Using /projects/ansible/ansible.cfg as config file

PLAYBOOK: site.yml *************************************************************
1 plays in site.yml

PLAY [deploy and configure site] ***********************************************

TASK [pkg : check if pkg tool needs bootstrapping] *****************************
task path: /projects/ansible/roles/pkg/tasks/main.yml:4
<mrbrown> ESTABLISH SSH CONNECTION FOR USER: ansible
<mrbrown> SSH: EXEC ssh -C -q -o PubkeyAuthentication=yes -o ControlMaster=auto -o ControlPersist=6000s -o StrictHostKeyChecking=no -o Port=2200 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/Users/dch/.ansible/cp/ansible-ssh-%h-%p-%r -tt mrbrown '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-gklmlnjqhzsyxvshemyjgwtkjdiolnzv; /usr/sbin/pkg -N'"'"'"'"'"'"'"'"' && sleep 0'"'"''
ok: [mrbrown] => {"changed": false, "invocation": {"module_args": {"_raw_params": "/usr/sbin/pkg -N"}, "module_name": "raw"}, "rc": 0, "stderr": "", "stdout": "pkg: 35 packages installed\r\n", "stdout_lines": ["pkg: 35 packages installed"]}

TASK [pkg : bootstrap pkg tool] ************************************************
task path: /projects/ansible/roles/pkg/tasks/main.yml:13
skipping: [mrbrown] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
```
##### ACTUAL RESULTS

``` shell
ansible-playbook site.yml --limit mrbrown\* -vvv                                                                              (1) (1s 1ms)
Using /projects//ansible/ansible.cfg as config file

PLAYBOOK: site.yml *************************************************************
1 plays in site.yml

PLAY [deploy and configure site] ***********************************************

TASK [pkg : check if pkg tool needs bootstrapping] *****************************
task path: /projects//ansible/roles/pkg/tasks/main.yml:4
<mrbrown> ESTABLISH SSH CONNECTION FOR USER: ansible
<mrbrown> SSH: EXEC ssh -q -o PubkeyAuthentication=yes -o ControlMaster=auto -o ControlPersist=6000s -o StrictHostKeyChecking=no -o Port=2200 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/Users/dch/.ansible/cp/ansible-ssh-%h-%p-%r -tt mrbrown 'sudo -H -S -n -u root '"'"'echo BECOME-SUCCESS-xoagidmebhurswchuqecjyudqqyvhnja; /usr/sbin/pkg -N'"'"''
fatal: [mrbrown]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"_raw_params": "/usr/sbin/pkg -N"}, "module_name": "raw"}, "rc": 1, "stderr": "", "stdout": "sudo: echo BECOME-SUCCESS-xoagidmebhurswchuqecjyudqqyvhnja; /usr/sbin/pkg -N: command not found\r\n", "stdout_lines": ["sudo: echo BECOME-SUCCESS-xoagidmebhurswchuqecjyudqqyvhnja; /usr/sbin/pkg -N: command not found"]}
...ignoring

TASK [pkg : bootstrap pkg tool] ************************************************
task path: /projects//ansible/roles/pkg/tasks/main.yml:13
<mrbrown> ESTABLISH SSH CONNECTION FOR USER: ansible
<mrbrown> SSH: EXEC ssh -q -o PubkeyAuthentication=yes -o ControlMaster=auto -o ControlPersist=6000s -o StrictHostKeyChecking=no -o Port=2200 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/Users/dch/.ansible/cp/ansible-ssh-%h-%p-%r -tt mrbrown 'sudo -H -S -n -u root '"'"'echo BECOME-SUCCESS-tdsvqapdrcixbrhyrhzxkvraroalxpeu; /usr/bin/env ASSUME_ALWAYS_YES=1 /usr/sbin/pkg bootstrap -f'"'"''
```
##### WORKAROUNDS
1. Uncommenting the `# executable=/bin/sh` as recommended by @bcoca works however the shell of the ansible user at the remote end is _already_ /bin/sh

The result then is: `{"changed": false, "invocation": {"module_args": {"_raw_params": "/usr/sbin/pkg -N", "executable": "/bin/sh"}`
1. Patch via @bcoca 

``` diff
diff --git a/lib/ansible/modules/core b/lib/ansible/modules/core
index e992cfd..1d0f408 160000
--- a/lib/ansible/modules/core
+++ b/lib/ansible/modules/core
@@ -1 +1 @@
-Subproject commit e992cfd47324b0acaee2d3df01a6da10189ee0d9
+Subproject commit 1d0f4088973afd6440c58d7f2d14a0e3663b5146
diff --git a/lib/ansible/playbook/play_context.py b/lib/ansible/playbook/play_context.py
index 1f0a04f..4e5649c 100644
--- a/lib/ansible/playbook/play_context.py
+++ b/lib/ansible/playbook/play_context.py
@@ -447,7 +447,7 @@ class PlayContext(Base):

         if self.become:

-            if executable is None:
+            if not executable:
                 executable = self.executable

             becomecmd   = None
```

also works. Many thanks!

**Comment by jjshoe** [comment]
I believe this is already done:

https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/play_context.py#L450

**Comment by bcoca** [comment]
fixed via 4acac55d233ac09745b56c8a577a2404019d6e9e
